src/context/WidgetSettingContext.tsx

```diff
@@ -0,0 +1,209 @@
+
+interface WidgetSession {
+  strategy: 'auto' | 'manual';
+  userId: string;
+  expireAt: number;
+  channelUrl?: string; // The channel is optional since it is dynamically generated from a manual strategy.
+  sessionToken?: string;
+}
+
+interface BotStyle {
+  theme: 'light' | 'dark';
+  accentColor: string;
+  botMessageBGColor: string;
+  autoOpen: boolean;
+}
+
+type Context = {
+  initialized: boolean;
+  botStyle: BotStyle;
+  widgetSession: WidgetSession | null;
+  initManualSession: (sdk: SendbirdChatWith<[GroupChannelModule]>) => void;
+};
+
+const WidgetSettingContext = createContext<Context | null>(null);
+export const WidgetSettingProvider = ({
+  children,
+}: React.PropsWithChildren) => {
+  const {
+    createGroupChannelParams,
+    firstMessageData,
+    userId: injectedUserId,
+    configureSession,
+    applicationId: _appId,
+    botId: _botId,
+    apiHost,
+  } = useConstantState();
+
+  const appId = _appId as string;
+  const botId = _botId as string;
+
+  if (!appId || !botId) {
+    throw new Error('applicationId or botId is not defined');
+  }
+
+  const sessionStrategy: 'auto' | 'manual' =
+    typeof configureSession === 'function' && !!injectedUserId
+      ? 'manual'
+      : 'auto';
+
+  const inProgress = React.useRef(false);
+  const [initialized, setInitialized] = useState(false);
+  const [botStyle, setBotStyle] = useState<BotStyle>({
+    theme: 'light',
+    accentColor: '#742DDD',
+    botMessageBGColor: '#EEEEEE',
+    autoOpen: false,
+  });
+  const [widgetSession, setWidgetSession] = useState<WidgetSession | null>(
+    null
+  );
+
+  async function initSessionByStrategy(strategy: 'auto' | 'manual') {
+    const cachedSession = getWidgetSessionCache({
+      appId,
+      botId,
+    });
+
+    // Reuse the session if it is valid.
+    if (
+      cachedSession &&
+      cachedSession.strategy === strategy &&
+      !isPastTime(cachedSession.expireAt)
+    ) {
+      const response = await getWidgetSetting({
+        host: apiHost,
+        appId,
+        botId,
+        createUserAndChannel: false,
+      });
+
+      setBotStyle(response.botStyle);
+      setWidgetSession({
+        strategy: sessionStrategy,
+        userId: cachedSession.userId,
+        expireAt: cachedSession.expireAt,
+        channelUrl: cachedSession.channelUrl,
+        sessionToken: cachedSession.sessionToken,
+      });
+    } else {
+      const response = await getWidgetSetting({
+        host: apiHost,
+        appId,
+        botId,
+        createUserAndChannel: strategy === 'auto',
+      });
+
+      setBotStyle(response.botStyle);
+
+      if (sessionStrategy === 'auto' && response.user && response.channel) {
+        const session = {
+          strategy: sessionStrategy,
+          expireAt: response.user.expireAt,
+          userId: response.user.userId,
+          sessionToken: response.user.sessionToken,
+          channelUrl: response.channel.channelUrl,
+        };
+        setWidgetSession(session);
```

Comment:
minor: I think we can take below code outside
```
const response = await getWidgetSetting({
        host: apiHost,
        appId,
        botId,
        createUserAndChannel: false,
      });

      setBotStyle(response.botStyle);
      setWidgetSession({
        strategy: sessionStrategy,
        userId: cachedSession.userId,
        expireAt: cachedSession.expireAt,
        channelUrl: cachedSession.channelUrl,
        sessionToken: cachedSession.sessionToken,
      });
```
Declare outsite, update inside.

Reply:
-> https://github.com/sendbird/chat-ai-widget/pull/238/commits/1ff08fc508338987935ae0d95a3deb51db8db40c